src/main/java/com/iota/iri/LedgerValidator.java

```diff
@@ -198,103 +187,89 @@ private void updateConsistentHashes(final Snapshot snapshot, Hash tip, int index
+            hasSnapshot = currentState != null && Snapshot.isConsistent(latestSnapshot.patch(currentState));
+            if (hasSnapshot) {
+                updateSnapshotMilestone(milestoneVM.getHash(), milestoneVM.index());
+                StateDiffViewModel stateDiffViewModel;
+                stateDiffViewModel = new StateDiffViewModel(currentState, milestoneVM.getHash());
+                if(currentState.size() != 0) {
+                    stateDiffViewModel.store(tangle);
                 }
+                latestSnapshot.merge(currentState, milestoneVM.index());
             }
-            return hasSnapshot;
         }
+        this.milestone.latestSnapshot.rwlock.writeLock().unlock();
+        return hasSnapshot;
     }
 
-    public boolean checkConsistency(final Snapshot snapshotReference, List<Hash> hashes) throws Exception {
-        Snapshot snapshot;
-        synchronized (snapshotReference.snapshotSyncObject) {
-            snapshot = new Snapshot(snapshotReference);
-        }
+    public boolean checkConsistency(List<Hash> hashes) throws Exception {
+        Set<Hash> visitedHashes = new HashSet<>();
+        Map<Hash, Long> diff = new HashMap<>();
+
+        milestone.latestSnapshot.rwlock.readLock().lock();
         for(Hash hash: hashes) {
-            if (!isTipConsistent(snapshot, hash)) return false;
+            if (!updateDiff(visitedHashes, diff, hash)) return false;
         }
+        milestone.latestSnapshot.rwlock.readLock().unlock();
         return true;
     }
 
-    public boolean isTipConsistent(Snapshot snapshot, Hash tip) throws Exception {
-        TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle, tip);
-        if(!transactionViewModel.isSolid()) {
+    public boolean updateDiff(Set<Hash> approvedHashes, final Map<Hash, Long> diff, Hash tip) throws Exception {
+        if(!TransactionViewModel.fromHash(tangle, tip).isSolid()) {
             return false;
         }
-        boolean isConsistent;
-        synchronized (snapshot.snapshotSyncObject) {
-            synchronized (snapshot.approvalsSyncObject) {
-                if (!(isConsistent = snapshot.approvedHashes.contains(tip))) {
-                    Hash tail = transactionViewModel.getHash();
-                    int latestSyncIndex = snapshot.index();
-                    Map<Hash, Long> currentState = getLatestDiff(snapshot, tail, latestSyncIndex, false);
-                    isConsistent = currentState != null && snapshot.patch(currentState, latestSyncIndex).isConsistent();
-                    if (isConsistent) {
-                        updateConsistentHashes(snapshot, tip, latestSyncIndex);
-                        snapshot.merge(snapshot.patch(currentState, latestSyncIndex));
-                    }
-                }
-            }
+        if (approvedHashes.contains(tip)) return true;
+        Set<Hash> visitedHashes = new HashSet<>(approvedHashes);
```

Comment:
can you explain why these are copied, and "copied" back (`approvedHashes.addAll(visitedHashes);`) at the end w/o any change?
or am I missing something?

Reply:
I was missing the change, it's fixed now